In [33]:
#import libraries
import pandas as pd
import sqlalchemy
import numpy as np

In [2]:
#create engine
engine = sqlalchemy.create_engine('mysql+pymysql://root:@localhost:3306/Lucid')

In [3]:
#read users table
users = pd.read_sql_table("users", engine, index_col='id')

In [4]:
#drop irrelevant columns
users.drop(['image','password','remember_token'],axis = 1, inplace=True)

In [5]:
#views users head
users.head()

name   username                         email provider  \
id                                                                            
1      Eniayomi Oluwaseyi   eniayomi   oluwaseyieniayomi@gmail.com   google   
2           Elijah Okokon    DMatrix           okoelijah@gmail.com   google   
3            Jeffrey Ogah  jeff.ogah           jeff.ogah@gmail.com   google   
4   Oluwaseyi Oluwapelumi     nathan     nathanoluwaseyi@gmail.com   google   
5                    PoRH      lamar  paulchibiukeigweze@gmail.com   google   

              provider_id          created_at          updated_at  \
id                                                                  
1   100892238782850407484 2019-08-01 18:57:22 2019-08-01 18:57:22   
2   105584000674868760138 2019-08-01 19:05:58 2019-08-01 19:05:58   
3   101235103722282698796 2019-08-01 21:55:39 2019-08-01 21:55:39   
4   107233455544779563919 2019-08-02 17:52:47 2019-08-02 17:52:47   
5   103431607062264005570 2019-08-03 12:33:51 2019-08-03 12:33:51   

                                            short_bio  
id                                                     
1                Software Developer | DevOPs Engineer  
2                                       Web Developer  
3   Front End Developer | React Developer | Mentor...  
4   | Software Developer | DevOps Engineer | @linu...  
5             I Am lamar and you don't think am real?

In [6]:
#read users post
posts = pd.read_sql_table('posts', engine,index_col='id')

In [7]:
#drop irrelevant columns
posts.drop(['status_id','action','post_id'], axis = 1, inplace=True)

In [8]:
#view posts head
posts.head()

user_id                              title  \
id                                               
1      2077  What i have learnt so far on HTML   
2      1719                   HTML BEGINS HERE   
4      1310            My Laziness In The Open   
6      1787                          MY TASK 2   
7       167                             Task 2   

                                              content        tags  \
id                                                                  
1   I learnt how to use the table tag as i have us...        None   
2    I am on this journey with start.ng, and here ...  Technology   
4   I have not been attending classes on the HNG c...        None   
6   My journey on **StartNG** pre-internship progr...        None   
7    A Summary on The “idongesit.html” CV, Its Str...        None   

                                     slug          created_at  \
id                                                              
1   what-i-have-learnt-so-far-on-html-985 2019-08-25 20:39:01   
2                    html-begins-here-4d6 2019-08-25 20:44:36   
4             my-laziness-in-the-open-029 2019-08-25 20:50:33   
6                           my-task-2-649 2019-08-25 20:51:36   
7                              task-2-7e5 2019-08-25 20:51:39   

            updated_at image  
id                            
1  2019-08-28 12:30:00     1  
2  2019-08-28 12:30:00     1  
4  2019-08-28 12:30:00     1  
6  2019-08-28 12:30:00     1  
7  2019-08-28 12:30:00     1

In [27]:
posts.user_id.nunique()

308

In [9]:
#read contact table
contact= pd.read_sql_table('contact_settings',engine,index_col='id')

In [10]:
#drop NaN
contact.dropna(axis = 1,inplace=True)

In [11]:
#view head
contact.head()

user_id                        email
id                                      
1        78  emmanuelezenwigbo@gmail.com
2        90           dward009@gmail.com
3        53            apeji93@gmail.com
4       133          ireabueke@gmail.com
5        92          bindas.fs@gmail.com

In [12]:
#read following table
following = pd.read_sql_table('following',engine)

In [13]:
following.head()

my_id  follower_id  status
0      3            6       1
1      6            3       1
2      3            2       1
3      3            7       1
4      7            2       1

In [68]:
#read notifications table
notifications = pd.read_sql_table('notifications',engine)

In [71]:
#drop NaN
notifications.dropna(axis = 1,inplace=True)
notifications.drop(['id'], axis = 1, inplace=True)

In [77]:
notifications.head()

sender_id  user_id  status     action       type          created_at  \
0          4        7       1   Followed  Following 2019-08-26 10:43:56   
1          1        4       1   Followed  Following 2019-08-26 10:48:21   
2          3       45       0   Followed  Following 2019-08-26 10:51:39   
3          1        4       1   Followed  Following 2019-08-26 12:32:46   
4          1        4       1  Commented       Post 2019-08-26 16:13:17   

           updated_at  
0 2019-08-26 19:20:28  
1 2019-08-26 14:05:42  
2 2019-08-26 10:51:39  
3 2019-08-26 14:05:42  
4 2019-08-26 16:13:25

In [78]:
notifications.user_id.nunique()

25

In [17]:
#read thoughts table
thoughts = pd.read_sql_table('thoughts',engine,index_col='id')

In [18]:
thoughts.head()

user_id                                            content  \
id                                                               
1      2195                                           hhhhhhhh   
2       148                      Can't believe I'm learning ML   
3      2212  - what you have learned so far\r\nI've learned...   
4         1                                         Hey people   
5      2223                                               nice   

            created_at          updated_at  
id                                          
1  2019-08-26 14:30:26 2019-08-26 14:30:26  
2  2019-08-26 14:57:35 2019-08-26 14:57:35  
3  2019-08-26 15:35:41 2019-08-26 15:35:41  
4  2019-08-26 16:12:07 2019-08-26 16:12:07  
5  2019-08-26 16:24:55 2019-08-26 16:24:55

In [19]:
#merging posts and contact table
df1 = pd.merge(posts,contact,on='user_id')
df1.head()

user_id               title  \
0      329  My Journey on HTML   
1      329         Please note   
2      329  My Journey on HTML   
3      329  My learning so far   
4      942                 why   

                                             content  tags  \
0  Using the Hyper Text Markup Language (HTML) ha...  None   
1  The screenshot of my HTML CV attached to my la...  None   
2  <p>Using the Hyper Text Markup Language (HTML)...         
3  <p>So far, have learnt how to optimally use Sl...         
4                                       lucid why na  None   

                     slug          created_at          updated_at  \
0  my-journey-on-html-ce5 2019-08-25 20:54:07 2019-08-28 12:30:34   
1         please-note-4f4 2019-08-25 21:06:38 2019-08-28 12:30:01   
2  my-journey-on-html-ce5 2019-08-27 12:14:33 2019-08-28 12:30:34   
3  my-learning-so-far-56f 2019-08-27 12:14:33 2019-08-28 12:30:34   
4                 why-8be 2019-08-25 21:04:09 2019-08-28 12:30:01   

                                 image                     email  
0                                    1       topejoan2@gmail.com  
1                                 None       topejoan2@gmail.com  
2  topejoan2/images/img-q8lju377v8.png       topejoan2@gmail.com  
3                                            topejoan2@gmail.com  
4                                 None  enoekanem.vibe@gmail.com

In [149]:
#merging posts and notifications
df2 = pd.merge(posts,notifications,on='user_id')
df2.drop(['tags','created_at_x','updated_at_x','created_at_y','updated_at_y'], axis = 1, inplace = True)

In [21]:
#merging posts and thoughts
df3 = pd.merge(posts,thoughts,on='user_id')
df3.head()

user_id                                     title  \
0     2195  CODING101: THROUGH THE EYE OF A BEGINNER   
1     2212                  WHAT I'VE LEARNED SO FAR   
2     2212                                USING HTML   
3        4                     Good Afternoon People   
4        4                     Good Afternoon People   

                                           content_x  tags  \
0  I was scrolling through my tweets the other da...  None   
1  **WHAT HAVE YOU LEARNED SO FAR?**\n\n  \n\nI'v...  None   
2  I joined a newbie and I'm really ecstatic to b...  None   
3                                                Hey  None   
4                                                Hey  None   

                                          slug        created_at_x  \
0  coding101-through-the-eye-of-a-beginner-6e6 2019-08-26 10:06:06   
1                  what-ive-learned-so-far-6d6 2019-08-26 15:41:21   
2                               using-html-dd2 2019-08-26 20:34:31   
3                    good-afternoon-people-288 2019-08-26 16:10:44   
4                    good-afternoon-people-288 2019-08-26 16:10:44   

         updated_at_x image  \
0 2019-08-28 12:30:07  None   
1 2019-08-28 12:30:12  None   
2 2019-08-28 12:30:17     1   
3 2019-08-28 12:30:12  None   
4 2019-08-28 12:30:12  None   

                                           content_y        created_at_y  \
0                                           hhhhhhhh 2019-08-26 14:30:26   
1  - what you have learned so far\r\nI've learned... 2019-08-26 15:35:41   
2  - what you have learned so far\r\nI've learned... 2019-08-26 15:35:41   
3                                         Hey people 2019-08-26 23:29:07   
4                                                Hey 2019-09-05 18:19:25   

         updated_at_y  
0 2019-08-26 14:30:26  
1 2019-08-26 15:35:41  
2 2019-08-26 15:35:41  
3 2019-08-26 23:29:07  
4 2019-09-05 18:19:25

In [22]:
#merge users and contact
df4 = pd.merge(users,contact, on='email')
df4.head()

name      username                   email provider  \
0     Lawal Idris   idriseun222   idriseun222@gmail.com   google   
1     Yossi Apeji       apeji93       apeji93@gmail.com   google   
2  wanapere akpos      akpos008      akpos008@gmail.com   google   
3      Gift Omodu   giftomodu20   giftomodu20@gmail.com   google   
4   Joseph Achonu  josephachonu  josephachonu@gmail.com   google   

             provider_id          created_at          updated_at  \
0  114996915816431931347 2019-08-11 12:36:15 2019-08-11 12:36:15   
1  111946827587303311452 2019-08-18 22:49:58 2019-08-18 22:49:58   
2  106544284168226714299 2019-08-18 22:50:35 2019-08-18 22:50:35   
3  111607729815913807483 2019-08-18 22:50:53 2019-08-18 22:50:53   
4  104643006509003406701 2019-08-18 22:51:31 2019-08-18 22:51:31   

                                           short_bio  user_id  
0                                  Software Engineer       19  
1  I am a solution's architect and Economist look...       53  
2                                Cool and easy going       61  
3  Back-end Developer...\r\n\r\nError\r\n#404\r\n...       62  
4  A Tech ethusiast and a strong believer in educ...       70

In [23]:
df2.groupby('title')['action'].count().sort_values(ascending = False).head()

title
CEO -- Story so Far                                             105
BUILD AROUND YOUR DISCOVERY MECHANISM                            42
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAILY THAN YOU DO     35
DON'T ASK WHY YOU WERE REJECTED                                  35
THE SUNFLOWER                                                    35
Name: action, dtype: int64

From above we can see that CEO -- Story so Far has the most actions i.e commented or followed 

In [125]:
users.head()

name   username                         email provider  \
id                                                                            
1      Eniayomi Oluwaseyi   eniayomi   oluwaseyieniayomi@gmail.com   google   
2           Elijah Okokon    DMatrix           okoelijah@gmail.com   google   
3            Jeffrey Ogah  jeff.ogah           jeff.ogah@gmail.com   google   
4   Oluwaseyi Oluwapelumi     nathan     nathanoluwaseyi@gmail.com   google   
5                    PoRH      lamar  paulchibiukeigweze@gmail.com   google   

              provider_id          created_at          updated_at  \
id                                                                  
1   100892238782850407484 2019-08-01 18:57:22 2019-08-01 18:57:22   
2   105584000674868760138 2019-08-01 19:05:58 2019-08-01 19:05:58   
3   101235103722282698796 2019-08-01 21:55:39 2019-08-01 21:55:39   
4   107233455544779563919 2019-08-02 17:52:47 2019-08-02 17:52:47   
5   103431607062264005570 2019-08-03 12:33:51 2019-08-03 12:33:51   

                                            short_bio  
id                                                     
1                Software Developer | DevOPs Engineer  
2                                       Web Developer  
3   Front End Developer | React Developer | Mentor...  
4   | Software Developer | DevOps Engineer | @linu...  
5             I Am lamar and you don't think am real?

In [126]:
posts.head()

user_id                              title  \
id                                               
1      2077  What i have learnt so far on HTML   
2      1719                   HTML BEGINS HERE   
4      1310            My Laziness In The Open   
6      1787                          MY TASK 2   
7       167                             Task 2   

                                              content        tags  \
id                                                                  
1   I learnt how to use the table tag as i have us...        None   
2    I am on this journey with start.ng, and here ...  Technology   
4   I have not been attending classes on the HNG c...        None   
6   My journey on **StartNG** pre-internship progr...        None   
7    A Summary on The “idongesit.html” CV, Its Str...        None   

                                     slug          created_at  \
id                                                              
1   what-i-have-learnt-so-far-on-html-985 2019-08-25 20:39:01   
2                    html-begins-here-4d6 2019-08-25 20:44:36   
4             my-laziness-in-the-open-029 2019-08-25 20:50:33   
6                           my-task-2-649 2019-08-25 20:51:36   
7                              task-2-7e5 2019-08-25 20:51:39   

            updated_at image  
id                            
1  2019-08-28 12:30:00     1  
2  2019-08-28 12:30:00     1  
4  2019-08-28 12:30:00     1  
6  2019-08-28 12:30:00     1  
7  2019-08-28 12:30:00     1

In [151]:
df2.drop(['content','slug','image'], axis = 1, inplace=True)

In [152]:
df2.head()

user_id                             title  sender_id  status     action  \
0     2191      What you have learned so far       2191       0  Commented   
1     2191      What you wish to still learn       2191       0  Commented   
2     2191  What you wish should be improved       2191       0  Commented   
3        4             Good Afternoon People          1       1   Followed   
4        4             Good Afternoon People          1       1   Followed   

        type  
0       Post  
1       Post  
2       Post  
3  Following  
4  Following

In [154]:
df2.action.unique()

array(['Commented', 'Followed', 'Like', 'Replied', 'Love'], dtype=object)

In [161]:
def status(text):
    if text == 'Commented':
        return 1.0
    elif text == 'Followed':
        return 2.0
    elif text == 'Like':
        return 3.0
    elif text == 'Replied':
        return 4.0
    elif text == 'Love':
        return 5.0

In [162]:
df2['rating'] = df2['action'].apply(status)

In [163]:
df2.head()

user_id                             title  sender_id  status     action  \
0     2191      What you have learned so far       2191       0  Commented   
1     2191      What you wish to still learn       2191       0  Commented   
2     2191  What you wish should be improved       2191       0  Commented   
3        4             Good Afternoon People          1       1   Followed   
4        4             Good Afternoon People          1       1   Followed   

        type  rating  
0       Post     1.0  
1       Post     1.0  
2       Post     1.0  
3  Following     2.0  
4  Following     2.0

In [189]:
#create a pivot table
df = df2.pivot_table(index = 'sender_id', columns = 'title', values = 'status', aggfunc='mean')

In [190]:
df.head()

title      A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAILY THAN YOU DO  \
sender_id                                                                 
1                                                   0.333333              
2                                                   0.888889              
3                                                   1.000000              
4                                                   0.888889              
6                                                        NaN              

title      ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?  Am ok  \
sender_id                                                           
1                                                      0.0    NaN   
2                                                      NaN    NaN   
3                                                      1.0    NaN   
4                                                      1.0    NaN   
6                                                      NaN    NaN   

title      BUILD AROUND YOUR DISCOVERY MECHANISM  CEO -- Story so Far  \
sender_id                                                               
1                                       0.200000                  NaN   
2                                       0.888889                  NaN   
3                                       1.000000                  0.0   
4                                       0.909091                  0.0   
6                                            NaN                  NaN   

title      Class Day 1  DO YOU NEED A CO-FOUNDER?  \
sender_id                                           
1                  NaN                        0.0   
2                  NaN                        NaN   
3                  0.0                        1.0   
4                  0.5                        1.0   
6                  1.0                        NaN   

title      DON'T ASK WHY YOU WERE REJECTED  EVALUATING STARTUP IDEAS  First  \
sender_id                                                                     
1                                 0.333333                       0.0    1.0   
2                                 0.888889                       NaN    NaN   
3                                 1.000000                       1.0    1.0   
4                                 0.888889                       1.0    NaN   
6                                      NaN                       NaN    NaN   

title      ...  asda  html test  jbhjkhkbb  ochem  second  third  \
sender_id  ...                                                     
1          ...   NaN   1.000000        NaN    NaN     1.0    1.0   
2          ...   NaN   0.888889        NaN    NaN     NaN    NaN   
3          ...   NaN   1.000000        0.0    NaN     1.0    1.0   
4          ...   NaN   0.857143        NaN    NaN     NaN    NaN   
6          ...   NaN        NaN        NaN    NaN     NaN    NaN   

title      this is a test post  this is mine  what have done so far on HTML  \
sender_id                                                                     
1                     1.000000      1.000000                            NaN   
2                     0.888889      0.888889                            NaN   
3                     1.000000      1.000000                            NaN   
4                     0.857143      0.857143                            NaN   
6                          NaN           NaN                            NaN   

title      with picture  
sender_id                
1              1.000000  
2              0.888889  
3              1.000000  
4              0.857143  
6                   NaN  

[5 rows x 55 columns]

In [191]:

x = df['CEO -- Story so Far']

In [192]:
#finding the correlation of other articles with CEO--story so far article
corr = df.corrwith(x)

In [193]:
corr

title
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAILY THAN YOU DO         NaN
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?                      NaN
Am ok                                                                NaN
BUILD AROUND YOUR DISCOVERY MECHANISM                                NaN
CEO -- Story so Far                                             1.000000
Class Day 1                                                     0.522233
DO YOU NEED A CO-FOUNDER?                                            NaN
DON'T ASK WHY YOU WERE REJECTED                                      NaN
EVALUATING STARTUP IDEAS                                             NaN
First                                                                NaN
Good Afternoon People                                                NaN
HNG Pre-internship                                                   NaN
HOW TO AVOID BEING A ONE-HIT WONDER                                  NaN
Hello New Followers                          

Intuitively, we can see that CEO-- story so far is correlated with the titles with 0.522233 correlation values

In [196]:
#correlating the columns using pearson formula
matrix_corr = df.corr(method='pearson')

In [197]:
matrix_corr.head()

title                                               A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAILY THAN YOU DO  \
title                                                                                                              
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAIL...                                           1.000000              
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?                                              0.955072              
Am ok                                                                                             NaN              
BUILD AROUND YOUR DISCOVERY MECHANISM                                                        0.995061              
CEO -- Story so Far                                                                               NaN              

title                                               ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?  \
title                                                                                                 
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAIL...                                         0.955072   
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?                                            1.000000   
Am ok                                                                                           NaN   
BUILD AROUND YOUR DISCOVERY MECHANISM                                                      0.984108   
CEO -- Story so Far                                                                             NaN   

title                                               Am ok  \
title                                                       
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAIL...    NaN   
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?       NaN   
Am ok                                                 NaN   
BUILD AROUND YOUR DISCOVERY MECHANISM                 NaN   
CEO -- Story so Far                                   NaN   

title                                               BUILD AROUND YOUR DISCOVERY MECHANISM  \
title                                                                                       
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAIL...                               0.995061   
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?                                  0.984108   
Am ok                                                                                 NaN   
BUILD AROUND YOUR DISCOVERY MECHANISM                                            1.000000   
CEO -- Story so Far                                                                   NaN   

title                                               CEO -- Story so Far  \
title                                                                     
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAIL...                  NaN   
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?                     NaN   
Am ok                                                               NaN   
BUILD AROUND YOUR DISCOVERY MECHANISM                               NaN   
CEO -- Story so Far                                                 1.0   

title                                               Class Day 1  \
title                                                             
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAIL...    -0.519170   
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?       -0.866025   
Am ok                                                       NaN   
BUILD AROUND YOUR DISCOVERY MECHANISM                 -0.520205   
CEO -- Story so Far                                    0.522233   

title                                               DO YOU NEED A CO-FOUNDER?  \
title                                                                           
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAIL...                   0.955072   
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?                      1.000000   
Am ok                                                            

In [199]:
#posts in which the sender 1 acted on by either liking, loving, commenting and so on
df.iloc[0].dropna()

title
A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAILY THAN YOU DO    0.333333
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?                 0.000000
BUILD AROUND YOUR DISCOVERY MECHANISM                           0.200000
DO YOU NEED A CO-FOUNDER?                                       0.000000
DON'T ASK WHY YOU WERE REJECTED                                 0.333333
EVALUATING STARTUP IDEAS                                        0.000000
First                                                           1.000000
Good Afternoon People                                           1.000000
HOW TO AVOID BEING A ONE-HIT WONDER                             0.000000
SHALLOW THINKING                                                0.000000
TEst                                                            1.000000
THE SUNFLOWER                                                   0.333333
TWO YEARS IN THE TRENCHES BUILDING HOTELS.NG                    0.000000
This is lucid                                

In [230]:
#creating a series for sender corr
sender_corr = pd.Series()

sender_id = 1

for post in df.iloc[sender_id-1].dropna().index:
    corr_list = matrix_corr[post].dropna()*df.iloc[sender_id-1][post]
    sender_corr = sender_corr.append(corr_list)

In [231]:
#grouping user correlations by the index and summing up
sender_corr = sender_corr.groupby(sender_corr.index).sum()

In [233]:
#create an empty list for posts
post_list = []

for i in range(len(df.iloc[sender_id-1].dropna().index)):
    if df.iloc[sender_id-1].dropna().index[i] in sender_corr:
        post_list.append(df.iloc[sender_id-1].dropna().index[i])
    else:
        pass
sender_corr = sender_corr.drop(post_list)

In [234]:
print('Hi, based on the post that you reacted to: \n')
for i in df.iloc[sender_id-1].dropna().index:
    print(i)
print('\n I would suggest you read these posts \n')
for i in sender_corr.sort_values(ascending= False).index[:5]:
    print(i)

Hi, based on the post that you reacted to: 

A CHINESE VILLAGER WHO SELLS MORE SOFTWARE DAILY THAN YOU DO
ARE YOU CAPABLE OF BEING RUTHLESS TO GET AHEAD?
BUILD AROUND YOUR DISCOVERY MECHANISM
DO YOU NEED A CO-FOUNDER?
DON'T ASK WHY YOU WERE REJECTED
EVALUATING STARTUP IDEAS
First
Good Afternoon People
HOW TO AVOID BEING A ONE-HIT WONDER
SHALLOW THINKING
TEst
THE SUNFLOWER
TWO YEARS IN THE TRENCHES BUILDING HOTELS.NG
This is lucid
Took a long while
Welcome to Lucid
YOUR HIGH IQ WILL KILL YOUR STARTUP
Your face PHP script
html test
second
third
this is a test post
this is mine
with picture

 I would suggest you read these posts 

Working round the clock on Lucid
Testing tags and really long posts
Testing image
Test
Stressful Day
